### PR TITLE
[WIP] Add `LAMMPSMDCalc` for molecular dynamics using LAMMPS

### PR DESCRIPTION
**Pull request by rul048**
## Summary

This PR provides a calculator with argument specification and default values compatible that are almost fully compatible with `MDCalc` using ASE for molecular dynamics (MD) simulation using LAMMPS.

Major changes:

- **Add `LAMMPSMDCalc` class**: Introduces a new `PropCalc` subclass to run molecular dynamics simulations via LAMMPS, supporting NVE, NVT, and NPT ensembles with Nose-Hoover thermostat.  
- **Template-driven input generation**: Inspired by pymatgen, creates a `lammps_templates/md.template` file to avoid hardcoding LAMMPS commands and make customization effortless. **As our collection of templates grows, we will provide a community-driven platform in which to use and share LAMMPS workflows for our MVL members, external contributors and users-fostering collaboration and ensuring that research can be easily reproduced.**
- **Energy Minimization**: Leverages the existing `RelaxCalc` to relax the atomic postions of the input structure before MD, ensuring consistency with `MDCalc`. Also provides the option to perform energy minimization using LAMMPS.
- **Trajectory and energy extraction**: Parses LAMMPS dump and log files to return ASE-compatible trajectories and compute average potential, kinetic, and total energies over the last N frames.  
- **Multiple GLIPs support**: Supports MatGL, CHGNet, MACE, ORB, MatterSim and FAIRChem. The`gnnp_type` for MatGL potentials will be automatically inferred.  
- **Unit tests**: Adds pytest tests covering energy outputs for different ensembles to ensure correctness.

Important Note: As model name is required in LAMMPS script, the calculator passed in the `LAMMPSMDCalc` must be a string.

## Todos

If this is work in progress, what else needs to be done?

- Support additional ensembles (e.g., Langevin, Andersen, Bussi, Berendsen). 
- Encourage MVL members to upload their templates to lammps_templates.
- Implement diffusion coefficient and implicit anharmonic phonon calculation based on MD simulations (eg. `DiffusionCalc`, `DynaPhononCalc` and `PheasyCalc`).  
- Update project documentation and examples to demonstrate MD workflows.

## Checklist

- [x] Google-format docstrings added (verify with `ruff`).  
- [x] Type annotations included and clean (verify with `mypy`).  
- [x] Tests added for all new features and edge cases.  
- [ ] Wherever applicable, add `@due.dcite` decorators to reference relevant papers.  

Tip: Install `pre-commit` hooks to auto-check types and linting before every commit:

```sh
pip install -U pre-commit
pre-commit install
```

### CODE REVIEW

**Comment by shyuep** [comment]
If the parameters are all the same, wouldn't it be easier to have a use_lammps option and you have a calc_ase, and calc_lammps method and calc calls either one depending on whether use_lammps is true? Both are afterall just MD calculations.

File: src/matcalc/_lammps.py [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+        traj = []
+        for snapshot in parse_lammps_dumps(self.trajfile):
+            lattice = snapshot.box.to_lattice()
+            coords = snapshot.data[["x", "y", "z"]].to_numpy()
+            species = snapshot.data["element"].tolist()
+            struct = Structure(lattice, species, coords)
+
+            traj.append(AseAtomsAdaptor.get_atoms(struct))
+        # Select the last 'self.frames' frames from the trajectory for further analysis.
+        traj = traj[-self.frames :]
```
Comment (atulcthakur) [inline src/matcalc/_lammps.py:237]: This will run out of memory and break if the trajectory gets too large. Alternative is to keep track of the frame number you are reading with `enumerate.` Since we know the total number of frames, it is easy to save only the last nframes.

File: src/matcalc/lammps_templates/md.template [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+thermo_style    custom step time cpu pe ke etotal temp press vol density  # control the thermo data type to output
+thermo          $loginterval  # output thermo data every $loginterval step(s)
+
+dump            1 all custom $loginterval $trajfile id element x y z  # dump a snapshot every $loginterval step(s)
+dump_modify     1 sort id element $symbol_set
+
+# Actions
+
+timestep        $timestep
+run             $steps
```
Comment (atulcthakur) [inline src/matcalc/lammps_templates/md.template:51]: Add a `write_data` and `write_restart` command at the end to save the last snapshot as a data and a restart file.

File: src/matcalc/lammps_templates/md.template [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+# Output settings
+
+thermo_style    custom step time cpu pe ke etotal temp press vol density  # control the thermo data type to output
+thermo          $loginterval  # output thermo data every $loginterval step(s)
+
+dump            1 all custom $loginterval $trajfile id element x y z  # dump a snapshot every $loginterval step(s)
+dump_modify     1 sort id element $symbol_set
+
+# Actions
```
Comment (atulcthakur) [inline src/matcalc/lammps_templates/md.template:48]: We should also add a line to save restart files every ~ `1000` frames or so, in order to allow for restarts if the run gets interrupted. You can even have lammps iterate back and forth between two file names if you are worried that there might be too many restart files.

**Review by atulcthakur — COMMENTED** [review]
Just a few simple comments.

**Comment by shyuep** [comment]
Look at the `_backend` class. That is how we are going to support different backends. Of course, there are some PropCalcs that may only support either ASE or LAMMPS. But for those where we can have common method APIs, we should just do that. You will need to implement the `run_lammps` method so that all relaxation and static calculations are supported. That will at least make most of the code work on both.

**Comment by rul048** [comment]
Hi Professor, now the PR works fine! Thank you so much for your guidance! Without your help, it would probably take me several more days to solve this problem. Always can learn a lot from you. @shyuep

**Comment by rul048** [comment]
Next, I will implement the `calc_lammps` as the backend and try to merge the `MDCalc` and `LAMMPSMDCalc` together. At least now, these calculators can work well independently.
